package/base-files/files/lib/functions/ipv4.sh

```diff
@@ -0,0 +1,166 @@
+uint_max=4294967295
+
+# Common error handler
+__die() {
+    echo "$1" >&2
+    exit 1
+}
+
+# check that $1 is only base 10 digits, and that it doesn't
+# exceed 2^32-1
+check_uint32() {
+    local n="$1"
+
+    [ -z "$n" -o -n "${n//[0-9]/}" ] && __die "Not a decimal integer ($n)"
+    [ "$n" -gt $uint_max ] && __die "Out of range ($n)"
+    [ "$((n + 0))" != "$n" ] && __die "Not normalized notation ($n)"
+
+    echo "$n"
+}
+
+# return a count of the number of bits set in $1
+bitcount() {
+    local c
+    c="$(check_uint32 "$1")" || exit 1
+
+    c=$((((c >> 1) & 0x55555555) + (c & 0x55555555)))
+    c=$((((c >> 2) & 0x33333333) + (c & 0x33333333)))
+    c=$((((c >> 4) & 0x0f0f0f0f) + (c & 0x0f0f0f0f)))
+    c=$((((c >> 8) & 0x00ff00ff) + (c & 0x00ff00ff)))
+    c=$((((c >> 16) & 0x0000ffff) + (c & 0x0000ffff)))
+
+    echo "$c"
+}
+
+# tedious but portable with busybox's limited shell
+# we check each octet to be in the range of 0..255,
+# and also make sure there's no extaneous characters.
+ip2int() {
+    local ip="$1" n ret=0
+
+    case "$ip" in
+    [0-9].*)
```

Comment:
This correctly parses address bytes in the octal notation but not hexadecimal - both have been supported by the version of ipcalc.sh prior to this (and are supported by tools like ping from iputils).

Reply:
It actually doesn't parse octal.  And no one has used octal dotted-quad notation in 40 years.

Reply:
My bad, I overlooked the [1-9] at the start.